titus-server-master/src/main/java/com/netflix/titus/master/scheduler/resourcecache/OpportunisticCpuAllocation.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.master.scheduler.resourcecache;
+
+import java.util.Objects;
+
+public class OpportunisticCpuAllocation {
+    private final String taskId;
+    private final String agentId;
+    private final String allocationId;
+    private final int cpuCount;
+
+    public OpportunisticCpuAllocation(String taskId, String agentId, String allocationId, int cpuCount) {
+        this.taskId = taskId;
+        this.agentId = agentId;
+        this.allocationId = allocationId;
+        this.cpuCount = cpuCount;
+    }
+
+    public String getTaskId() {
+        return taskId;
+    }
+
+    public String getAgentId() {
+        return agentId;
+    }
+
+    public String getAllocationId() {
+        return allocationId;
+    }
+
+    public int getCpuCount() {
+        return cpuCount;
+    }
+
+    @Override
+    public String toString() {
+        return "OpportunisticCpuAllocation{" +
+                "taskId='" + taskId + '\'' +
+                ", agentId='" + agentId + '\'' +
+                ", allocationId='" + allocationId + '\'' +
+                ", cpuCount=" + cpuCount +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof OpportunisticCpuAllocation)) {
+            return false;
+        }
+        OpportunisticCpuAllocation that = (OpportunisticCpuAllocation) o;
+        return cpuCount == that.cpuCount &&
+                taskId.equals(that.taskId) &&
+                agentId.equals(that.agentId) &&
+                allocationId.equals(that.allocationId);
+    }
+
+    public static OpportunisticCpuAllocationBuilder newBuilder() {
+        return new OpportunisticCpuAllocationBuilder();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(taskId, agentId, allocationId, cpuCount);
+    }
+
+    public static final class OpportunisticCpuAllocationBuilder {
+        private String taskId;
+        private String agentId;
+        private String allocationId;
+        private int cpuCount;
+
+        private OpportunisticCpuAllocationBuilder() {
+        }
+
+        public OpportunisticCpuAllocationBuilder withTaskId(String taskId) {
+            this.taskId = taskId;
+            return this;
+        }
+
+        public OpportunisticCpuAllocationBuilder withAgentId(String agentId) {
+            this.agentId = agentId;
+            return this;
+        }
+
+        public OpportunisticCpuAllocationBuilder withAllocationId(String allocationId) {
+            this.allocationId = allocationId;
+            return this;
+        }
+
+        public OpportunisticCpuAllocationBuilder withCpuCount(int cpuCount) {
+            this.cpuCount = cpuCount;
+            return this;
+        }
+
+        public OpportunisticCpuAllocation build() {
```

Comment:
Check arguments to eliminate null values.

Reply:
I'm not opposed, but tried to be consistent, we don't do this in any other `Builder` I could find